File: src/main/java/aoc2019/Day10.java
package aoc2019;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/*
  * Day 10: Monitoring Station
  * https://adventofcode.com/2019/day/10
  */
public class Day10 {

    static final String[] MAP1 = { ".#..#", ".....", "#####", "....#", "...##" };

    static final String[] MAP2 = { "......#.#.", "#..#.#....", "..#######.", ".#.#.###..", ".#..#.....", "..#....#.#",
            "#..#....#.", ".##.#..###", "##...#..#.", ".#....####" };
    static final String[] MAP3 =
            { ".#....#####...#..", "##...##.#####..##", "##...#...#.#####.", "..#.....X...###..", "..#.#.....#....##" };
    

    static class Asteroid extends Point {

        static final int MAX_ANGLE = (int) ((2 * Math.PI) * 1000);
        
        Asteroid(int x, int y) {
            super(x,y);
        }

        /**
         * angle from this to a in milli-rads, counted clockwise, North=0.
         * this function is as f***ed up as its name indicates.
         */
        int mradClockwiseFromNorth(Asteroid a) {
        	// standard atan2 counts counter-clockwise from East=0
            var rad = Math.atan2(-(a.x - x), a.y - y) + Math.PI; // clockwise, rotate 90° ccw
            return (int) (1000 * rad) % MAX_ANGLE;
        }
    }
    
    public static void main(String[] args) throws Exception {
        var map = Util.linesArray("input10.txt");
        List<Asteroid> asteroids = readMap(map, '#');
        
        System.out.println("=== part1 ===");
        Asteroid loc = new Day10().part1(asteroids);

        System.out.println("=== part2 ===");
        new Day10().part2(asteroids, loc);
    }
    

    static List<Asteroid> readMap(String[] map, char marker) {
        ArrayList<Asteroid> asteroids = new ArrayList<>();
        for (int y = 0; y < map.length; ++y)
            for (int x = 0; x < map[y].length(); ++x)
                if (map[y].charAt(x) == marker)
                    asteroids.add(new Asteroid(x, y));
        return asteroids;
    }


    Asteroid part1(List<Asteroid> asteroids) {        
        Map<Asteroid, Long> result = new HashMap<>();

        for (Asteroid station : asteroids) {         
        	// for all other asteroids ...
            var targets = asteroids.stream().filter(a -> a != station);
            // ... determine the angle to station, for each angle value only one asteroid is visible.
            var visible = targets.map(station::mradClockwiseFromNorth).distinct().count();
            result.put(station, visible);
        }

        // find the station with max visible targets
        var bestLocation = result.entrySet().stream().max(Comparator.comparing(e -> e.getValue())).orElse(null);
        System.out.println(bestLocation);
        return bestLocation.getKey();
    }

    
    void part2(List<Asteroid> asteroids, Asteroid station) {
        asteroids.remove(station);
        var vaporized = new ArrayList<Asteroid>();

        while (!asteroids.isEmpty()) {
        	// calculate and sort the angles, starting from north=0 (clockwise)
            int[] angles = asteroids.stream().mapToInt(station::mradClockwiseFromNorth).distinct().sorted().toArray();
            for (int angle : angles) {
            	// for each angle, zap the nearest asteroid
                Asteroid firstHit = asteroids.stream()
                        .filter(a -> station.mradClockwiseFromNorth(a) == angle)
                        .min(Comparator.comparing(station::manhattan))
                        .orElse(null);
                
                vaporized.add(firstHit);
            }
            asteroids.removeAll(vaporized);
        } 

        System.out.println(vaporized.get(199));
    }

}

File: src/main/go/aoc2019/day10.go
package main

import (
	"fmt"
	"math"
	"sort"
	"time"
)

// Run10 Day 10: Monitoring Station
func Run10() {

	t := time.Now()
	fmt.Println("--- part 1")
	day10Part1()

	fmt.Println("--- part 2")
	day10Part2()

	fmt.Printf("=== end (%d ms) ===\n", time.Now().Sub(t).Milliseconds())
}

func day10Part1() {
	input := ReadInputStrings("input10.txt")
	m := make(map[Point]bool)
	for y, s := range input {
		for x, c := range s {
			if c == '#' {
				m[Point{x, y}] = true
			}
		}
	}

	angle := func(p1, p2 Point) int {
		//return int(10000 * math.Atan2(float64(p2.y-p1.y), float64(p2.x-p1.x)))
		return int(10000 * (math.Atan2(-float64(p2.x-p1.x), float64(p2.y-p1.y)) + math.Pi))
	}
	max := 0
	best := Point{0, 0}
	for s := range m {
		vis := make(map[int]int)
		for a := range m {
			an := angle(s, a)
			_, ex := vis[an]
			if !ex {
				vis[an] = 1
			} else {
				vis[an]++
			}
		}
		// fmt.Printf("(%d,%d) sees %d\n", s.x, s.y, len(vis))
		if len(vis) > max {
			max = len(vis)
			best = s
		}
	}

	fmt.Println(max, " from ", best)
}

func day10Part2() {
	input := ReadInputStrings("input10.txt")
	m := make(map[Point]bool)
	for y, s := range input {
		for x, c := range s {
			if c == '#' {
				m[Point{x, y}] = true
			}
		}
	}

	station := Point{20, 21}
	angle := func(p1, p2 Point) int {
		return int(10000 * (math.Atan2(-float64(p2.x-p1.x), float64(p2.y-p1.y)) + math.Pi))
	}

	vis := make(map[int][]Point)
	for a := range m {
		alpha := angle(station, a)
		_, ex := vis[alpha]
		if !ex {
			vis[alpha] = []Point{a}
		} else {
			vis[alpha] = append(vis[alpha], a)
		}
	}

	// sort angles
	keys := make([]int, 0, len(vis))
	for k := range vis {
		keys = append(keys, k)
	}
	sort.Slice(keys, func(i, j int) bool { return keys[i] < keys[j] })
	// fmt.Println(keys)

	// sort asteroids by distance per angle
	for a := range vis {
		sort.Slice(vis[a], func(i, j int) bool { return vis[a][i].mdist(station) < vis[a][j].mdist(station) })
	}
	// fmt.Println(vis[0])

	// zap asteroids, nearest first, by increasing angle
	for i, zapped := 0, 0; zapped <= 200; i = (i + 1) % len(keys) {
		alpha := keys[i]
		if _, ex := vis[alpha]; ex {
			zapped++
			fmt.Println("kill ", zapped, ": ", vis[alpha][0])
			vis[alpha] = vis[alpha][1:]
		}

	}
}
